File: java/src/test/java/it/giannibombelli/wsc2026/booking/application/usecases/BookingRejectingTest.java
package it.giannibombelli.wsc2026.booking.application.usecases;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.booking.infrastructure.SqliteBookingRepository;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.testsupport.AggregateFactory;
import it.giannibombelli.wsc2026.testsupport.AggregateFactory;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;

import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.RejectBooking;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class BookingRejectingTest {
    private BookingRepository repository;
    private BookingRejecting rejection;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("booking");
        repository = new SqliteBookingRepository(dataSource);
        rejection = new BookingRejecting(repository, event -> {
        });
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new BookingRejecting(null, event -> {
            }))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullCommand() {
            assertThatThrownBy(() -> rejection.invoke(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldReturnBookingRejected() {
            Booking booking = AggregateFactory.createBooking();
            repository.save(booking);
            Money amount = new Money(new BigDecimal("75.00"));

            var event = rejection.invoke(new RejectBooking(booking.id(), booking.giftCardReference(), amount));

            assertThat(event.aggregateId()).isEqualTo(booking.id());
            assertThat(event.giftCardReference()).isEqualTo(booking.giftCardReference().value());
            assertThat(event.amount()).isEqualTo(amount);
        }

        @Test
        void shouldFailIfBookingNotFound() {
            var nonExistingId = EntityId.generate(it.giannibombelli.wsc2026.booking.domain.booking.BookingId::new);

            assertThatThrownBy(() -> rejection.invoke(new RejectBooking(nonExistingId, AggregateFactory.createBooking().giftCardReference(), new Money(new BigDecimal("10.00")))))
                .isInstanceOf(IllegalStateException.class);
        }
    }
}

File: typescript/tests/booking/application/usecases/bookingRejecting.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { rejectBooking } from '@/booking/application/commands/bookingConfirmationCommands.js';
import { BookingRejecting } from '@/booking/application/usecases/bookingRejecting.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import { SqliteBookingRepository } from '@/booking/infrastructure/sqliteBookingRepository.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';
import { BookingAggregateFactory } from '../../../testsupport/booking/aggregateFactory.js';
import type { BookingEvent } from '@/booking/domain/events/bookingEvent.js';
import type { EventPublisher } from '@/common/application/events/eventPublisher.js';

describe('BookingRejecting', () => {
  let repository: SqliteBookingRepository;
  let rejection: BookingRejecting;

  const noOpPublisher: EventPublisher<BookingEvent> = {
    publish: () => {},
  };

  beforeEach(() => {
    const database = DatabaseSetup.initializeInMemoryDb('booking');
    repository = new SqliteBookingRepository(database);
    rejection = new BookingRejecting(repository, noOpPublisher);
  });

  describe('construction', () => {
    it('reject null repository', () => {
      expect(() => new BookingRejecting(null as unknown as SqliteBookingRepository, noOpPublisher)).toThrow();
    });
  });

  describe('execution', () => {
    it('reject null command', () => {
      expect(() => rejection.invoke(null as unknown as ReturnType<typeof rejectBooking>)).toThrow();
    });

    it('should return booking rejected', () => {
      const booking = BookingAggregateFactory.createBooking();
      repository.save(booking);
      const giftCardId = generateId((value) => new GiftCardId(value));
      const amount = new Money(75);

      const event = rejection.invoke(rejectBooking(booking.id(), giftCardId, amount));

      expect(event.aggregateId).toEqual(booking.id());
      expect(event.giftCardId).toEqual(giftCardId);
      expect(event.amount).toEqual(amount);
    });

    it('should fail if booking not found', () => {
      const nonExistingId = generateId((value) => new BookingId(value));
      const giftCardId = generateId((value) => new GiftCardId(value));

      expect(() => rejection.invoke(rejectBooking(nonExistingId, giftCardId, new Money(10)))).toThrow();
    });
  });
});
